orchagent/orch.cpp

```diff
@@ -239,23 +247,67 @@ void ConsumerBase::dumpPendingTasks(vector<string> &ts)
 
 void Consumer::execute()
 {
-    // ConsumerBase::execute_impl<swss::ConsumerTableBase>();
     SWSS_LOG_ENTER();
 
-    auto table = static_cast<swss::ConsumerTableBase *>(getSelectable());
-    std::deque<KeyOpFieldsValuesTuple> entries;
-    table->pops(entries);
+    size_t total_size = 0;
+
+    while (true)
+    {
+        size_t popped_size = 0; // number of entries popped from the redis table
+
+        auto entries = std::make_shared<std::deque<KeyOpFieldsValuesTuple>>();
+        getConsumerTable()->pops(*entries);
 
-    // add to sync
-    addToSync(entries);
+        popped_size = entries->size();
+        total_size += popped_size;
 
-    drain();
+        pushRingBuffer([=](){
+            addToSync(entries);
+        });
+
+        if (!gBatchSize || popped_size * 10 <= (size_t)gBatchSize || total_size >= ROUTE_SYNC_PPL_SIZE) {
+            // some program doesn't initialize gBatchSize and use TableConsumable::DEFAULT_POP_BATCH_SIZE instead
+            break;
+        }
+    }
+
+    pushRingBuffer([=](){
+        drain();
+    });
+}
+
+void Executor::pushRingBuffer(AnyTask&& task)
+{
+    if (!gRingBuffer || !gRingBuffer->threadCreated) 
+    {
+        // execute the task right now in this thread if gRingBuffer is not initialized
+        // or the ring thread is not created, or this executor is not served by gRingBuffer
+        task();
+    }
+    else if (!gRingBuffer->Serves(getName())) // not served by ring thread
+    {
+        while (!gRingBuffer->IsEmpty() || !gRingBuffer->Idle) {
+            gRingBuffer->notify();
+            std::this_thread::sleep_for(std::chrono::milliseconds(SLEEP_MSECONDS));
```

Comment:
Every consumer in orchagent work in same thread. which means the code here will sleep entire orchagent 0.5seconds.

I'm warry about 0.5seconds delay for critical event handling.

Reply:
In this case, the consumer of APP_ROUTE_TABLE is actively working in the ring thread...so the consumer in the main thread waits until the ring thread pauses, since we don't want the two threads to work at the same time (because the two threads are currently dealing with different consumers). Actually not the entire orchagent sleeps for 0.5s since route table consumer is still running.

In the original single-thread case, all consumers work in the main thread and execute one by one, the later consumer executes after the previous consumer...so the 0.5 seconds is reasonable, i think

Reply:
I noticed in the Orch::addExecutor(Executor* executor) the APP_ROUTE_TABLE_NAME been add to both m_consumerMap and gRingBuffer, is that by design?

And that make APP_ROUTE_TABLE consumer shared in 2 threads. but sonic-swss-common does not support multiple thread and code may crash.

Reply:
yes, but  APP_ROUTE_TABLE is added to gRingBuffer just for housekeeping, so that gRingBuffer->Serves() could determine whether a table is served by ring.

we don't need sonic-swss-common to support multi-threading, because in our design, when a route table consumer is running, the main thread only performs pops, and addToSync() and drain() are done in ring thread, the two threads don't share common data, they can work in parallel. And only the route table consumer has multi-threading. For other consumers, since the main thread would wait for the ring to pause, and they never push tasks to ring thread, everything is only done in the main thread, it's still single-thread here.

Reply:
I understand the async task for addToSync and drain are running in ring buffer thread in OrchDaemon::popRingBuffer().

However, the sleep 0.5seconds is running in main thread.

If main thread need wait ring buffer finish, then why not wait by signal or mutex?

If the task in ring buffer take < 0.5seconds, this PR actually make orchagent slow.

If the task in ring buffer take > 0.5seconds, then there some code may running in both ring buffer thread and main thread and cause race conduction issue, because the drain() method will call orch->dotask(), in some orch, dotask() method will call method defined in other orch.

Reply:
For more information, the consumer of APP_ROUTE_TABLE_NAME will call RouteOrch::DoTask()

    const int routeorch_pri = 5;
    vector<table_name_with_pri_t> route_tables = {
        { APP_ROUTE_TABLE_NAME,        routeorch_pri },
        { APP_LABEL_ROUTE_TABLE_NAME,  routeorch_pri }
    };
    gRouteOrch = new RouteOrch(m_applDb, route_tables, gSwitchOrch, gNeighOrch, gIntfsOrch, vrf_orch, gFgNhgOrch, gSrv6Orch);

However:
1. the consumer of APP_LABEL_ROUTE_TABLE_NAME, will run in main thread, it also will call RouteOrch::DoTask().
2. In the code of RouteOrch::DoTask(), it will call method of gSwitchOrch, gNeighOrch, gIntfsOrch, vrf_orch, gFgNhgOrch, gSrv6Orch, which means there are at least 6 other consumers running in main thread will have potential race condition issue with ring buffer.

Reply:
> I understand the async task for addToSync and drain are running in ring buffer thread in OrchDaemon::popRingBuffer().
> 
> However, the sleep 0.5seconds is running in main thread.
> 
> If main thread need wait ring buffer finish, then why not wait by signal or mutex?
> 
> If the task in ring buffer take < 0.5seconds, this PR actually make orchagent slow.
> 
> If the task in ring buffer take > 0.5seconds, then there some code may running in both ring buffer thread and main thread and cause race conduction issue, because the drain() method will call orch->dotask(), in some orch, dotask() method will call method defined in other orch.

1. if < 0.5 sec, since main thread already finishes pops when the ring is running, it may not be slow
2. if > 0.5 sec, no race condition, since main thread would not call doTask when ring is active
3. we can use mutex, although no lock already achieves correctness

Reply:
Let's user gSwitchOrch as example, gSwitchOrch own a consumer for APP_SWITCH_TABLE_NAME.

This table does not serve by ring buffer. my question is:

The drain() method of the APP_SWITCH_TABLE_NAME consumer is running in main thread or in ring buffer thread?

According to the code here, when this consumer execute(), the code here will:
1. notify ring buffer.
2. wait 0.5 seconds
3. run the drain() method at line 279. which is in main thread.

Then what will happen of the async task in ring buffer thread run longer than 0.5seconds?

Reply:
currently, only route table consumer has the chance to run two threads at the same time (and the two threads don't have thread issue, because the main thread only performs pops, which only modifies redis table and the local variable), for other consumers, when they are chosen to execute(), they can only do pops, and proceed only if the ring thread is not active, otherwise  wait for the ring thread to pause (to avoid thread issue). so that we maintain the order among consumers, the multi-threading is only within the route table consumer itself.

how to make sure the ring thread has paused?
- it's empty && idle flag is true (since even if the ring is empty, maybe it's still running the last popped task, so need to check the idle flag)

to make sure we don't change the sequence order, we need to make sure the ring buffer is empty before allowing the next consumer to do addToSync and drain.

to avoid the ring thread being stuck, when the main thread determines that it needs to wait, it not only waits for 0.5s, but also sends out cv.notify to make sure there is no deadlock.

To answer your question,
1. all tables other than route table, runs in main thread
2. when it's selected to execute(), it immediately performs pops, without waiting. In original design, this pops happens after the previous consumer's drain, but currently they can be parallel
3. when async task in ring runs > 0.5 sec, main thread checks whether ring thread is paused, and found that idle is false, so it notifies the ring and wait another 0.5s
4. after the 0.5s, the main thread checks whether ring thread is paused again
    - you may question, maybe 0.5s is too much, but for the main thread, compared with the original design, even if it wastes the 0.5s here, it saves the time for pops, they can be balanced out  
5. it can continue to do addToSync and drain only after the ring buffer is empty and the ring thread is not running

Reply:
I still don't understand:
1. for this: "when it's selected to execute(), it immediately performs pops, without waiting. In original design, this pops happens after the previous consumer's drain, but currently they can be parallel"

For other tables, the async task of drain still running in main thread, the main thread still needs wait the async task finish to continue, which means it's not parallel.


2. "Not 6 seconds, still 1 sec"

For every event, after pop, the main thread will sleep o.5 seconds, so if there are 10 event, the main thread sleep time will be 5 seconds. why still 1 sec?

Reply:
the wait is in a while loop, the wait is not once, if ring thread is still active, APP_SWITCH_TABLE_NAME consumer  keeps waiting

Reply:
Oh, sorry, my mistake.
But then do you mean when APP_SWITCH_TABLE_NAME consumer running in background, the main thread actually waiting in the while loop?

Also, the second point means, when an APP_SWITCH_TABLE_NAME events happen with any another table event, the second event will always block 0.5seconds.

Reply:
the 0.5 sec delay would only be possible if APP_SWITCH_TABLE_NAME follows APP_ROUTE_TABLE_NAME, if it follows other kinds of table, the condition ```(!gRingBuffer->IsEmpty() || !gRingBuffer->Idle)``` is 100% false

Reply:
Is there test result show how much performance improve? because the only improve is APP_ROUTE_TABLE_NAME table pops data parallel with drain data. and pops data actually very fast.

Not sure any critical event follows after an APP_ROUTE_TABLE_NAME event will have 0.5seconds delay is acceptable.

Reply:
it shows benefits in the scaled case, when pops is really slow (redis i/o)
https://github.com/sonic-net/SONiC/pull/1521 shows it helps to accelerate BGP loading by 50%

Reply:
#Closed